src/gg.regression.unity.bots/Runtime/Scripts/StateRecorder/KeyboardEventSender.cs

```diff
@@ -12,95 +13,102 @@ namespace RegressionGames.StateRecorder
-                            deltaStateEvent = null;
-                            keys.Clear();
-                        }
-                    }
-                }
-                if (!deltaStateEvent.HasValue)
-                {
-                    DeltaStateEvent.From(keyboard, out var newEvent);
-                    deltaStateEvent = newEvent;
                 }
+                
+                InputSystem.QueueEvent(eventPtr);
+            }
+        }
 
-                logList.Add(valueTuple);
-                keys.Add(key);
-                var inputControl = keyboard.allControls
-                    .FirstOrDefault(a => a is KeyControl kc && kc.keyCode == key) ?? keyboard.anyKey;
-                if (inputControl != null)
+        private static void QueueTextEvent(int replaySegment, Key key)
+        {
+            // send a text event so that 'onChange' text events fire
+            // convert key to text
+            if (KeyboardInputActionObserver.KeyboardKeyToValueMap.TryGetValue(key, out var possibleValues))
+            {
+                var value = _isShiftDown ? possibleValues.Item2 : possibleValues.Item1;
+                if (value == 0x00)
                 {
-                    inputControl.WriteValueIntoEvent(upOrDown == KeyState.Down ? 1f : 0f, deltaStateEvent.Value);
-
-                    if (upOrDown == KeyState.Down)
-                    {
-                        // send a text event so that 'onChange' text events fire
-                        // convert key to text
-                        if (KeyboardInputActionObserver.KeyboardKeyToValueMap.TryGetValue(((KeyControl)inputControl).keyCode, out var possibleValues))
-                        {
-                            var value = _isShiftDown ? possibleValues.Item2 : possibleValues.Item1;
-                            if (value == 0x00)
-                            {
-                                RGDebug.LogError($"Found null value for keyboard input {key}");
-                                return;
-                            }
-
-                            var inputEvent = TextEvent.Create(Keyboard.current.deviceId, value, time);
-                            RGDebug.LogInfo($"({replaySegment}) Sending Text Event for char: '{value}'");
-                            InputSystem.QueueEvent(ref inputEvent);
-                        }
-                    }
+                    RGDebug.LogError($"Found null value for keyboard input {key}");
+                    return;
                 }
 
+                var inputEvent = TextEvent.Create(Keyboard.current.deviceId, value, InputState.currentTime);
+                RGDebug.LogInfo($"({replaySegment}) Sending Text Event for char: '{value}'");
+                InputSystem.QueueEvent(ref inputEvent);
+            }
+        }
+
+        /**
+         * Allows updating multiple keys in the same event. 
+         */
+        public static void SendKeysInOneEvent(int replaySegment, List<(Key, KeyState)> keyStates)
```

Comment:
I thought this was going to accept a `Dictionary` instead of a list.

Right now, when I pass it a list of (A-UP, W-UP, W-DOWN, D-DOWN)... the A-UP and W-UP get thrown away and not sent.

It looks like we either need to make this take in a dictionary... or add code to detect when we already have the specific key in the dictionary during the for loop and send the existing event and start a new one for the rest of the keys

Reply:
@RG-nAmKcAz I think I'd prefer to convert this to a dictionary. If the application wants to do something over multiple frames, I think it would be better to have it handle that in its own logic. Otherwise we'll have to do something complicated with queuing up key state changes until after the next input system update and it could get messy accounting for the different ways the APIs could be used.

Reply:
I've made the following changes:
1. I've converted SendKeysInOneEvent to take a dictionary: this method now serves to just create a single event for the given set of keys.
2. In order to handle the case you mentioned above, I've introduced a coroutine called `KeyboardEventSender.SendKeyEvents` and updated the sample exploration MonoBehaviour to use it instead. It will do the auto-splitting of events as done previously, except that it will actually wait until the next Input System update before queuing up a new event, since otherwise the Input System will just flush all the events right away.

Note that beyond some local testing I am not actually able to test the entire MonoBehaviour currently, so you may want to test it and feel free to push any other changes as needed to this branch.